[RFC] eval: Change the point at which arg_errmsg and its length are changed

ZyX-I opened the pull request:
Ref #6437

Vim had the similar thing, but it was removed to make code clearer. Apparently it was wrong decision.

ZyX-I committed:
eval: Change the point at which arg_errmsg and its length are changed

Ref #6437

justinmk commented on src/nvim/eval.c:
```diff
@@ -13503,14 +13497,12 @@ static void f_resolve(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     EMSG2(_(e_listarg), "reverse()");
   } else if ((l = argvars[0].vval.v_list) != NULL
-             && !tv_check_lock(l->lv_lock, arg_errmsg, arg_errmsg_len)) {
+             && !tv_check_lock(l->lv_lock, N_("insert() argument"),
```
Should it be reverse()?

tweekmonster commented:
Does this test need to be updated?  https://github.com/neovim/neovim/blob/dd391bfca1f37093ba556f5da6a7f3eb81147fc0/test/unit/eval/typval_spec.lua#L2437-L2454

tweekmonster commented on src/nvim/eval/typval.c:
```diff
@@ -2042,11 +2042,14 @@ bool tv_islocked(const typval_T *const tv)
+/// @param[in]  name_len  Variable name length. Use #TV_TRANSLATE to translate
+///                       variable name and compute the length. Use #TV_CSTRING
+///                       to compute the length with strlen() without
+///                       translating.
```
Maybe clarify more by saying that constants are used to defer gettext translation and length calculation for performance since the name is only used for error messages.

ZyX-I committed:
eval: Add comment regarding why special values are needed

ZyX-I committed:
unittests: Add a test for TV_CSTRING

Not using enum{} because SIZE_MAX exceeds integer and I do not really like how
enum definition is described in C99:

1. Even though all values must fit into the chosen type (6.7.2.2, p 4) the type
   to choose is still implementation-defined.
2. 6.4.4.3 explicitly states that “an identifier declared as an enumeration
   constant has type `int`”. So it looks like “no matter what type was chosen
   for enumeration, constants will be integers”. Yet the following simple
   program:

        #include <stdint.h>
        #include <stdio.h>
        #include <stddef.h>

        enum { X=SIZE_MAX };

        int main(int argc, char **argv)
        {
          printf("x:%zu m:%zu t:%zu v:%zu",
                 sizeof(X), sizeof(SIZE_MAX), sizeof(size_t), (size_t)X);
        }

    yields one of the following using different compilers:

    - clang/gcc/pathcc: `x:8 m:8 t:8 v:18446744073709551615`
    - pcc/tcc: `x:4 m:8 t:8 v:1844674407370955161`

    If I remove the cast of X to size_t then pcc/tcc both yield `x:4 m:8 t:8
    v:4294967295`, other compilers’ output does not change.

    All compilers were called with `$compiler -std=c99 -xc -` (feeding program
    from echo), except for `tcc` which has missing `-std=c99`. `pcc` seems to
    ignore the argument though: it is perfectly fine with `-std=c1000`.

ZyX-I commented:
`.2` job hang at `[ RUN      ] TermClose event reports the correct <abuf>: `. Restarting.

justinmk commented:
Cherry-picked 39a9760 onto c60e409, compiled with `make CMAKE_BUILD_TYPE=RelWithDebInfo`. This removes a few strlen calls, but otherwise the results are almost identical. See updated comment: https://github.com/neovim/neovim/issues/6437#issuecomment-293876362

justinmk commented:
After doing some sampling profiling it looks like `tv_clear` is a big part of the problem. `_()` was added there, removing that translation helps.

justinmk commented on src/nvim/eval.c:
```diff
@@ -8152,7 +8150,6 @@ static void f_expand(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 static void f_extend(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 {
   const char *const arg_errmsg = N_("extend() argument");
```
Why can't this be cached? Changing locale during runtime should not be supported.

justinmk commented:
@ZyX-I Instead of these special-case TV_FOO things, what do you think about having some function like `vim_gettext` that internally keeps a hashmap, or something like that? And that map would be cleared if user does `:lang`.

tweekmonster commented:
> internally keeps a hashmap

Wouldn't there still be some overhead from computing the key and scanning the map?  Using the `TV_*` constants completely avoids unnecessary work.

jamessan commented:
> Using the TV_*  constants completely avoids unnecessary work.

Except for calling into gettext every time the translation is needed.  The question is which is worse -- calling gettext or looking up a value in a hash?

justinmk commented:
Hashmap will be quite fast, we could even key it by error number (integer) to avoid the hashing step.

Reason I'm proposing this is, though this PR does save a few thousand calls (out of ~1 million), it doesn't give even 1% speedup on the test case I'm using (sourcing sexp.vim). So I sympathize with @ZyX-I's effort to make the code clearer. With this PR the clarity is lost and the gain may not be worth it, but there could be other hotspots that aren't being provoked in the test case I've been using.

So this PR may still be worth having, but if we could make the mechanism somewhat generalized, we could avoid the `TV_*` magic numbers which aren't really related to the general problem of "avoid gettext".

tweekmonster commented:
> Except for calling into gettext every time the translation is needed.

I think it'll be better to cache it regardless.  I was commenting about the removal of `TV_*` in favor of the said cache.

> The question is which is worse -- calling gettext or looking up a value in a hash?

Looking up a hash value for something that's unused often 😬

tweekmonster commented:
> Reason I'm proposing this is, though this PR does save a few thousand calls (out of ~1 million), it doesn't give even 1% speedup on the test case I'm using (sourcing sexp.vim).

Ah, I didn't know about that part of the reasoning.

jamessan commented:
On a slight tangent, it'd be interesting if we could have performance data tracked over time and some sort of PR check that performance doesn't regress more than X%.  Taking a quick look at the GitHub integrations, nothing jumps out as offering this sort of check though.

tweekmonster commented:
If the `TV_*` constants are removed, is the length argument even necessary?  Just about every use of `tv_check_lock()` I can find is using `STRLEN()`.

justinmk commented:
We could track performance data in a JSON file similar to how the clint check works.

ZyX-I commented:
@tweekmonster This is a part of an ongoing effort to remove temporary NUL sets. `name_len` is used in some places like `set_var` and `do_unlet_var` (basically functions which are doing the parsing and execution).

ZyX-I committed:
eval/typval: Do not translate tv_clear argument, this is useless

ZyX-I commented:
@justinmk Actually in the current state that `encode_vim_to_nothing()` argument is not used anywhere, only `encode()` functions are using that. I will just remove the `_()`.

---

@justinmk By the way, I guess I know how to have (almost) minimal performance overhead for GETTEXT_CACHED: create per-message `static size_t errmsg_idx`, if it is zero set it to `gettext_last_idx++;`, store everything in a simple array in place of hash and store in that array two pointers: original text and translated text. If array in question is *not* allocated (and it could be not since we can easily get a number of error messages which are using this variant at compile time) could not think of anything faster (except for storing translated messages locally in a static variable, having a macros which will save a static variable address in the hash and hoping the compiler optimizes local static’s better then previous variant or you would not note the difference). The only question is how to hide errmsg_idx (even worse if you want to store translated messages).

Additionally `:language` may trigger recomputing all cached translations which is going to be slow, but slow only at `:language` call which is rare. Since sometimes it is needed to have determenistic errors via calling `:language[ messages] C` and restoring locale afterwards and such a case will yield a performance problem in this case there is another optimization: `:language[ messages] C` does not recompute anything, it saves what the previous locale was and sets a flag that non-translated text should be used, on next `lang ` call it is checked whether previous locale matches new and nothing is done in the case. This is also a possible optimization for a hash lookup if that “minimal overhead” variant is not OK.

Though the first optimization with static variables (both variants) looks like overoptimization in the current state, it is something to consider only if we will run into some problems with hash.

ZyX-I committed:
eval/encode: Do translate “… argument” strings, but only in conv_error

ZyX-I commented:
This should be ready again.

ZyX-I committed:
Merge branch 'master' into lazier-arg_errmsg-gettext
